backend/src/apiserver/server/run_artifact_server.go

```diff
@@ -0,0 +1,148 @@
+	missingParamErrorMessage = "missing path parameter: '%s'"
+)
+
+type RunArtifactServer struct {
+	resourceManager *resource.ResourceManager
+}
+
+// StreamArtifactV1 is an artifact streaming endpoint that streams artifacts directly from object storage
+// to the HTTP response without buffering the entire content in memory.
+// No size limits are imposed - the streaming approach itself provides the security benefit.
+func (s *RunArtifactServer) StreamArtifactV1(response http.ResponseWriter, r *http.Request) {
+	glog.Infof("Stream artifact v1 called")
+
+	vars := mux.Vars(r)
+
+	runID, ok := vars[RunKey]
+
+	if !ok {
+		s.writeErrorToResponse(response, http.StatusBadRequest, fmt.Errorf(missingParamErrorMessage, RunKey))
+		return
+	}
+
+	nodeID, ok := vars[NodeKey]
+	if !ok {
+		s.writeErrorToResponse(response, http.StatusBadRequest, fmt.Errorf(missingParamErrorMessage, NodeKey))
+		return
+	}
+
+	artifactName, ok := vars[ArtifactNameKey]
+	if !ok {
+		s.writeErrorToResponse(response, http.StatusBadRequest, fmt.Errorf(missingParamErrorMessage, ArtifactNameKey))
+		return
+	}
+
+	artifactFileExists, err := s.artifactFileExists(r.Context(), runID, nodeID, artifactName)
+	if err != nil {
+		s.writeErrorToResponse(response, http.StatusInternalServerError, err)
+		return
+	} else if !artifactFileExists {
+		s.writeErrorToResponse(response, http.StatusNotFound, fmt.Errorf("artifact not found: %v", err))
+		return
+	}
+
+	response.Header().Set("Content-Type", "application/octet-stream")
+	response.Header().Set("Cache-Control", "no-cache, private")
+	response.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", artifactName))
+	response.WriteHeader(http.StatusOK)
+
+	err = s.resourceManager.StreamArtifact(r.Context(), runID, nodeID, artifactName, response)
+	if err != nil {
+		glog.Errorf("Failed to stream artifact: %v", err)
+		// Since we've already started writing the response, we can't change the status code
+		// Just log the error and close the connection
+		return
+	}
+}
+
+func (s *RunArtifactServer) artifactFileExists(ctx context.Context, runID string, nodeID string, artifactName string) (bool, error) {
+	artifactPath, err := s.resourceManager.ResolveArtifactPath(runID, nodeID, artifactName)
+	if err != nil {
+		if isNotFoundError(err) {
+			return false, nil
+		} else {
+			return false, err
+		}
+	}
+	reader, err := s.resourceManager.ObjectStore().GetFileReader(ctx, artifactPath)
+	if err != nil {
+		if isNotFoundError(err) {
+			return false, nil
+		} else {
+			return false, err
+		}
+	}
+	defer func() {
+		if closeErr := reader.Close(); closeErr != nil {
+			glog.Warningf("Failed to close artifact reader: %v", closeErr)
+		}
+	}()
+	return true, nil
+}
+
+// StreamArtifact handles v2beta1 artifact streaming (same implementation as v1)
+func (s *RunArtifactServer) StreamArtifact(w http.ResponseWriter, r *http.Request) {
+	glog.Infof("Stream artifact v2 called")
+	s.StreamArtifactV1(w, r)
```

Comment:
nit: it should be the other way around, `StreamArtifactV1` should use the V2 code, since we will remove v1 api endpoints soon

Reply:
Done.